src/main/java/com/coreos/jetcd/LeaseImpl.java

```diff
@@ -62,158 +65,192 @@
+  }
+
+  private synchronized void removeKeepAlive(long leaseId) {
+    this.keepAlives.remove(leaseId);
+  }
+
+  private void closeKeepAlives() {
+    final LeaseKeepAliveResponseWithError errResp = new LeaseKeepAliveResponseWithError(
+        new IllegalStateException("Lease client has closed"));
+    this.keepAlives.values().forEach(ka -> {
+      ka.sentKeepAliveResp(errResp);
+      ka.close();
+    });
+    this.keepAlives.clear();
+  }
+
+  private void start() {
+    this.sendKeepAliveExecutor();
+    this.deadLineExecutor();
+  }
+
+  private void reset() {
+    this.keepAliveFuture.cancel(true);
+    this.keepAliveRequestStreamObserver.onCompleted();
+    this.keepAliveResponseStreamObserver.onCompleted();
+    this.sendKeepAliveExecutor();
+  }
+
+  private void sendKeepAliveExecutor() {
+    this.keepAliveResponseStreamObserver = this.createResponseStreamObserver();
+    StreamObserver<LeaseKeepAliveRequest> requestStreamObserver = this.leaseStub
+        .leaseKeepAlive(this.keepAliveResponseStreamObserver);
+    this.keepAliveRequestStreamObserver = requestStreamObserver;
+    this.keepAliveFuture = scheduledExecutorService
+        .scheduleAtFixedRate(() -> {
+          long now = System.currentTimeMillis();
+
+          // send keep alive req to the leases whose next keep alive is before now.
+          this.keepAlives.entrySet().stream()
+              .filter(entry -> entry.getValue().getNextKeepAlive() < now)
+              .map((Entry::getKey))
+              .collect(Collectors.toList())
+              .forEach(leaseID -> requestStreamObserver
+                  .onNext(this.newKeepAliveRequest(leaseID)));
+        }, 0, 500, TimeUnit.MILLISECONDS);
+  }
+
+  private StreamObserver<LeaseKeepAliveResponse> createResponseStreamObserver() {
+    return new StreamObserver<LeaseKeepAliveResponse>() {
+      @Override
+      public void onNext(LeaseKeepAliveResponse leaseKeepAliveResponse) {
+        processKeepAliveResponse(leaseKeepAliveResponse);
+      }
+
+      @Override
+      public void onError(Throwable throwable) {
+        processOnError();
+      }
+
+      @Override
+      public void onCompleted() {
+      }
+    };
+  }
+
+  private synchronized void processOnError() {
+    if (this.closed) {
+      return;
+    }
+
+    this.scheduledExecutorService.schedule(() -> reset(), 500, TimeUnit.MILLISECONDS);
+  }
+
+  private synchronized void processKeepAliveResponse(
+      LeaseKeepAliveResponse leaseKeepAliveResponse) {
+    if (this.closed) {
+      return;
+    }
+
+    long leaseID = leaseKeepAliveResponse.getID();
+    long ttl = leaseKeepAliveResponse.getTTL();
+
+    KeepAlive ka = this.keepAlives.get(leaseID);
+    if (ka == null) {
+      return;
+    }
+
+    if (ttl <= 0) {
+      // lease expired; close all keep alive
+      this.removeKeepAlive(leaseID);
+      ka.sentKeepAliveResp(new LeaseKeepAliveResponseWithError(
+          new IllegalStateException("Lease " + leaseID + " expired")));
```

Comment:
I need to rework the exception. right now it is just a placeholder for unchecked exception.
I took a look of spanner api. they just have one SpannerException with error code and error msg. 
https://github.com/GoogleCloudPlatform/google-cloud-java/blob/master/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerException.java

I probably will copy that structure but i need to think through how to handle with error code.